### ISSUE TITLE
problem with zarr files

### ISSUE THREAD

**Issue by fipoucat** [r-spatial/stars#564]
I am trying to replicate the example in stars package on mac but getting errors:

> dsn = 'ZARR:"/vsicurl/https://ncsa.osn.xsede.org/Pangeo/pangeo-forge/gpcp-feedstock/gpcp.zarr"'
> bounds = c(longitude = "lon_bounds", latitude = "lat_bounds")
> r = read_mdim(dsn, bounds = bounds)
Error in read_mdim(file, array_name, options) : 
  CHAR() can only be applied to a 'CHARSXP', not a 'NULL'
In addition: Warning messages:
1: In read_mdim(file, array_name, options) :
  GDAL Error 1: Decompressor blosc not handled
2: In read_mdim(file, array_name, options) :
  GDAL Error 1: Decompressor blosc not handled
3: In read_mdim(file, array_name, options) :
  GDAL Error 1: Decompressor blosc not handled
4: In read_mdim(file, array_name, options) :
  GDAL Error 1: Decompressor blosc not handled
5: In read_mdim(file, array_name, options) :
  GDAL Error 1: Decompressor blosc not handled
6: In read_mdim(file, array_name, options) :
  GDAL Error 1: Decompressor blosc not handled
7: In read_mdim(file, array_name, options) :
  GDAL Error 1: Decompressor blosc not handled

**Comment by edzer** [comment]
Thanks for trying! This indicates that you are trying to load a zarr file that has BLOSC compression, but your GDAL version used to compile your sf package was not compiled with BLOSC support. We're trying to get this included in binary CRAN distributions (win/macos) but it might take a while. You'd have to compile from source against a GDAL with BLOSC support. This is all very recent technology!

**Comment by fipoucat** [comment]
OK, thank you

**Comment by pepijn-devries** [comment]
Hi there,

Just wondering, as I'm experiencing the same issue, is there any news on getting BLOSC support from the CRAN distribution? It would make accessing ZARR files from R so much easier... For completeness sake I've included my session info for the session in which I reproduced the issue:

```
R version 4.3.2 (2023-10-31 ucrt)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 19045)

Matrix products: default


locale:
[1] LC_COLLATE=Dutch_Netherlands.utf8  LC_CTYPE=Dutch_Netherlands.utf8   
[3] LC_MONETARY=Dutch_Netherlands.utf8 LC_NUMERIC=C                      
[5] LC_TIME=Dutch_Netherlands.utf8    

time zone: Europe/Amsterdam
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] stars_0.6-4 sf_1.0-15   abind_1.4-5

loaded via a namespace (and not attached):
 [1] vctrs_0.6.5        terra_1.7-65       cli_3.6.2          rlang_1.1.2       
 [5] DBI_1.2.0          KernSmooth_2.23-22 generics_0.1.3     glue_1.6.2        
 [9] e1071_1.7-14       fansi_1.0.6        grid_4.3.2         classInt_0.4-10   
[13] tibble_3.2.1       lifecycle_1.0.4    compiler_4.3.2     dplyr_1.1.4       
[17] codetools_0.2-19   sessioninfo_1.2.2  Rcpp_1.0.11        pkgconfig_2.0.3   
[21] rstudioapi_0.15.0  R6_2.5.1           class_7.3-22       tidyselect_1.2.0  
[25] utf8_1.2.4         pillar_1.9.0       parallel_4.3.2     magrittr_2.0.3    
[29] tools_4.3.2        proxy_0.4-27       units_0.8-5  
```

Cheers,

Pepijn

**Comment by edzer** [comment]
Your setup should have blosc support.

**Comment by pepijn-devries** [comment]
Thank you for the quick response, and good to know that it should work. I will do some additional checks, to see if I'm doing something wrong. If not, I will try to formulate a simple reprex. I will report my findings here, or open a new issue if it is unrelated to the issue reported here

**Comment by edzer** [comment]
#663

**Comment by pepijn-devries** [comment]
https://r-spatial.org/r/2022/09/13/zarr.html now states

> update users have reported [failure](https://github.com/r-spatial/stars/issues/564) to reproduce the steps in this blog post. Note that this blog post was run on a version of sf linked to GDAL 3.4.3 from the [ubuntugis-unstable](https://github.com/r-spatial/sf#ubuntu) ppa. If you have sf linked to an older version of GDAL, or if the GDAL version was not linked against BLOSC, then this (or parts of it) may not work. Note that Zarr, and Zarr support in GDAL are all very recent developments, and it takes time for modern library support to propagate through e.g. CRAN build toolchains – we are working on it!

This is at least informative, but is there a way to check if an installed version of `sf` supports BLOSC? This way I could start developing routines for ZARR, and warn users if their installation does not support this feature...

**Comment by edzer** [comment]
> is there a way to check if an installed version of sf supports BLOSC?

I don't think there is a run-time check. Only if GDAL has a run-time check, we could make one. I only know of the list of supported drivers (`sf::st_drivers()`), not of supported compressors.

**Comment by pepijn-devries** [comment]
After having looked through GDAL documentation, I could not find any mention of this. So I don't think it is possible to test this once GDAL is compiled. I could of course just catch the error that is produced and go from there.

**Comment by pepijn-devries** [comment]
> Yes, or look for symbols in statically linked code, or for blosc in dynamically linked upstream libs, e.g.
> 
> > any(grepl("blosc", system(paste("ldd", system.file("libs/sf.so", package = "sf")), intern = TRUE)))
> [1] TRUE
> which is ugly too, and in addition platform dependent.

Yes, this approach doesn't work directly on Windows. But with a work-around via Git Bash I was able to call `ldd` on `sf.dll` and it doesn't list blosc. This makes sense since blosc doesn't work on my machine. However, it also doesn't list gdal, so I'm not sure how complete this list is. Anyway, I still don't get why sf doesn't build with blosc support on my machine. I tried to build from source with Rtools 4.3, which is bundled with the blosc static library (I checked). It does build `sf` without any problems, but without bosc...
